### PR TITLE
Non unitary method for Linear Qu Differential Equations

### PR DESCRIPTION
**Pull request by dgan181**
Changes made in the unitary method and non unitary method added. 
@Roger-luo can you take a look at the non unitary code, I am unable to figure where I am mistaken.

### CODE REVIEW

File: src/QuLDE.jl [hunk 1]
```diff
@@ -70,19 +151,56 @@ function quldecircuit(blk::QuLDEParam, M::Matrix)
...
     x = prob.u0
-    T = Int(log2(k+1))
-
-    blk = QuLDEParam(k,t)
-    cir = quldecircuit(blk,M)
-
-    inreg = ArrayReg(x) ⊗ zero_state(T) ⊗ ( (blk.C_tilda/blk.N) * zero_state(1) )+  ArrayReg(b) ⊗ zero_state(T) ⊗ ((blk.D_tilda/blk.N) *ArrayReg(bit"1") )
+    Mu = M/opnorm(M)
+    siz, = size(M)
+    nbit = Int(log2(siz))
```
Comment (Roger-luo) [inline src/QuLDE.jl]: We have a `log2i` function implemented for integers in `BitBasis`.

File: test/QuLDE_tests.jl [hunk 2]
```diff
@@ -33,4 +33,15 @@ end
...
+    qprob = QuLDEProblem(An, b, x, tspan)
+    f(u,p,t) = An*u + b;
+    prob = ODEProblem(f, x, tspan)
+
+    sol = solve(prob, Tsit5(), dt = 0.1, adaptive = :false)
+    s = vcat(sol.u...)
+
+    out = solve(qprob, QuLDE(), k)
+
+    @test isapprox.(s[end-1:end], out, atol = 0.1) |> all
```
Comment (Roger-luo) [inline test/QuLDE_tests.jl]: sorry I need to catch up a little bit, which paper are you implementing in this PR?

File: src/QuLDE.jl [hunk 3]
```diff
@@ -46,12 +47,92 @@ struct QuLDEParam{W,L,Q, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixB
...
+        C_tilda = C_tilda + C(0)
+        N = sqrt(C_tilda^2 + D_tilda^2)
+        V = [C_tilda D_tilda; D_tilda -1*C_tilda]/N
+        V = convert(Array{ComplexF64,2},V)
+        V = matblock(V)
+        #tested
+
+        col1 = zeros(2^k,1)
+        col2 = zeros(2^k,1)
+        temp = map(0:k) do j 2^k - 2^(k-j) + 1 end
```
Comment (Roger-luo) [inline src/QuLDE.jl]: It would be better to use 0:k in the for loop and move this lambda there.

File: src/QuLDE.jl [hunk 4]
```diff
@@ -46,12 +47,92 @@ struct QuLDEParam{W,L,Q, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixB
...
+        cont = 0
+        for i in temp
```
Comment (Roger-luo) [inline src/QuLDE.jl]: e.g

```julia
for j in 0:k
     i = 2^k - 2^(k-j) + 1
     col1[i] = sqrt(C(cont))
     # blabla
end
```

you don't want to allocate this unnecessary array here.

File: src/QuLDE.jl [hunk 5]
```diff
@@ -46,12 +47,92 @@ struct QuLDEParam{W,L,Q, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixB
...
+            col1[i] = sqrt(C(cont))
+            cont += 1
+        end
+        cont = 1
+        for i in temp[2:end]
```
Comment (Roger-luo) [inline src/QuLDE.jl]: same issue

File: src/QuLDE.jl [hunk 6]
```diff
@@ -46,12 +47,92 @@ struct QuLDEParam{W,L,Q, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixB
...
+
+        C_tilda = 0
+        D_tilda = 0
+        for i in 1:k
+            C_tilda = C_tilda + C(i)
+            D_tilda = D_tilda + D(i)
+        end
+        C_tilda = C_tilda + C(0)
+        N = sqrt(C_tilda^2 + D_tilda^2)
+        V = [C_tilda D_tilda; D_tilda -1*C_tilda]/N
```
Comment (Roger-luo) [inline src/QuLDE.jl]: Why not use `ComplexF64[C_tilda D_tilda ... ]` directly? There is an extra conversion happens

File: src/QuLDE.jl [hunk 7]
```diff
@@ -46,12 +47,92 @@ struct QuLDEParam{W,L,Q, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixB
     end
 end
 
-function quldecircuit(blk::QuLDEParam, M::Matrix)
-    T = Int(log2(blk.k+1))
-    siz, = size(M)
-    nbit = Int(log2(siz))
+struct QuLDEnonUnitParam{W,L,Q,NL,Nbit, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixBlock{Q}, LType <: GeneralMatrixBlock{NL}}
+    k::Int
```
Comment (Roger-luo) [inline src/QuLDE.jl]: can you leave some comment on which is which for each member here? It is really hard to read.

File: test/QuLDE_tests.jl (hunk 2 above)
Comment (dgan181) [reply test/QuLDE_tests.jl]: https://arxiv.org/abs/1807.04553.
Appendix A , A is non-Unitary :

File: README.md [hunk 8]
```diff
@@ -5,3 +5,34 @@
...
+
+- Quantum Algorithms for Linear Differential Equations
+  - Based on Taylor Expansion (ref : arxiv.org/abs/1807.04553)
+  - Based on HHL (ref : arxiv.org/abs/1010.2745v2)
+
+
+## Built With
+
+* [Yao](https://github.com/QuantumBFS/Yao.jl) - A framework for Quantum Algorithm Design
+* [QuAlgorithmZoo](https://github.com/QuantumBFS/QuAlgorithmZoo.jl) - A repository for Quantum Algorithms
```
Comment (Roger-luo) [inline README.md]: Do we still have this as dependency?

Comment (dgan181) [reply README.md]: No. I added that because I have code from the repository (HHL, QFT and PE blocks).

**Comment by Roger-luo** [comment]
It looks good to me in general. But I'd like to make it more general later (e.g so it works for cuda). I'll merge this in the morning.

**Comment by Roger-luo** [comment]
Sorry for the delay. I just finished Yao's paper roughly today.

**Comment by dgan181** [comment]
No worries. Let me know how I can improve it.

File: src/QuLDE.jl [hunk 9]
```diff
@@ -41,17 +58,86 @@ struct QuLDEParam{W,L,Q, VType <: GeneralMatrixBlock{1}, SType <: GeneralMatrixB
...
+            D_tilda = D_tilda + D(i)
+        end
+        C_tilda = C_tilda + C(0)
+        N = sqrt(C_tilda+ D_tilda)
+        C_tilda = sqrt(C_tilda)
+        D_tilda = sqrt(D_tilda)
+        V = ComplexF64[C_tilda D_tilda; D_tilda -1*C_tilda]/N
+        V = matblock(V)
+        #tested
+        VS1 = rand(ComplexF64,2^k,2^k)
```
Comment (Roger-luo) [inline src/QuLDE.jl]: so this is an example that it is not generic enough to let it run on CUDA. But let's do it later, make things work first. And you could also keep this in mind the following coding.

To make it generic enough for CUDA, we need at least to support a general floating point precision (which means the program can also use `Float32`)

Comment (dgan181) [reply src/QuLDE.jl]: Is there an example I take a look at, to sort of better understand how to go about it?

Comment (Roger-luo) [reply src/QuLDE.jl]: For example, you could just let all `ComplexF64` be a type parameter (if you read the implementation of YaoBlocks, we never use a specific type, so the code would be very generic).

**Pull request merged**